highlight/highlight
# Send clientID from client.js

**UPDATE:** This PR now only contains the client changes. See description of all changes below.

---

Linear Issue: https://linear.app/highlight/issue/HIG-2384/backfill-unauthenticated-sessions-for-a-given-user

---

* Adds a `client_id` column to `sessions` and logic for setting it on `initializeSession` calls.
* Adds an `identified` column to `sessions` and logic for setting it on `identifySession` calls.
* Updates the client to send a `client_id` param in `initializeSession`. Also stores the `client_id` in `localStorage` so it can be persisted across sessions in a browser.
* Adds logic to `identifySession` to query past sessions where `identified = false` and identify them.

Note that the backfilling logic will only be enabled for the Highlight project to start. We will remove this in a follow up PR.

These changes will be broken up into smaller PRs so they can be deployed independently:

1. `backend/` + public graph changes
1. `client/` changes
1. `frontend/` changes

## Testing Locally

* Check out this branch and view the app. This will make a call to `initializeSession` that will have a `client_id` param sent with it.
* Create some session records to backfill in your DB by:
  1. Go in your BD and assign that `client_id` to other sessions that have no `user_properties` set. 
  3. Copy the value, hop into an incognito window and set that value in your `localStorage` (`localStorage.setItem("highlightClientID", "your-client-id")`) then visit the app, but don't log in.
* Inspect your DB to ensure you have sessions with a matching `client_id` where `identified = false` and there are no `user_properties`.
* Go back to your original browser window and trigger an `identifySession` call.
* Ensure the previously unidentified sessions are now identified.
